src/core/lib/promise/mpsc.cc

```diff
@@ -0,0 +1,419 @@
+              std::memory_order_relaxed)) {
+        return false;
+      }
+    } else {
+      // Observed active tokens less than max_queued_.
+      // This must continue for the duration of this code path since it is the
+      // only incrementer.
+      return true;
+    }
+  }
+}
+
+void Mpsc::DrainMpscFrom(bool wakeup, Node* node) {
+  GRPC_LATENT_SEE_INNER_SCOPE("Mpsc::DrainMpscFrom");
+  while (node != nullptr) {
+    auto next =
+        node->next_.exchange(Node::kWakerPtr, std::memory_order_acquire);
+    if (next & Node::kWakerPtr) {
+      DCHECK_NE(next, Node::kWakerPtr);
+      Waker* waker = reinterpret_cast<Waker*>(next & ~Node::kWakerPtr);
+      if (wakeup) waker->Wakeup();
+      delete waker;
+      next = 0;
+    }
+    if (node != &stub_) ReleaseTokensAndClose(wakeup, node);
+    node = reinterpret_cast<Node*>(next);
+  }
+}
+
+Poll<Mpsc::Node*> Mpsc::Dequeue() {
+  GRPC_LATENT_SEE_INNER_SCOPE("Mpsc::Dequeue");
+  Node* tail = tail_;
+  uintptr_t next = tail->next_.load(std::memory_order_acquire);
+retry_all:
+  if (tail == &stub_) {
+    if (next == 0) {
+      // List is (ephemerally) empty - create a waker so we get woken up
+      // when the queue is not empty.
+      Waker* waker = new Waker(GetContext<Activity>()->MakeNonOwningWaker());
+      if (!tail->next_.compare_exchange_weak(
+              next, reinterpret_cast<uintptr_t>(waker) | Node::kWakerPtr,
+              std::memory_order_acq_rel)) {
+        delete waker;
+        goto retry_all;
+      }
+      return Pending{};  // pending
+    }
+    if (next & Node::kWakerPtr) {
+      // null next waker => list closed
+      DCHECK_NE(next, Node::kWakerPtr);
+      // List is (ephemerally) empty - but we've already asked to be notified
+      // when non-empty.
+      return Pending{};  // pending
+    }
+    tail = reinterpret_cast<Node*>(next);
+    tail_ = tail;
+    next = tail->next_.load(std::memory_order_acquire);
+  }
+  if (next != 0 && (next & Node::kWakerPtr) == 0) {
+    tail_ = reinterpret_cast<Node*>(next);
+    return tail;
+  }
+  Node* head = head_.load(std::memory_order_acquire);
+  // Only the reader can close, so we should not see that here.
+  DCHECK_NE(head, nullptr);
+  if (tail != head) {
+    auto tail_next = head->next_.load(std::memory_order_acquire);
+    while (true) {
+      if (tail_next != 0 && (tail_next & Node::kWakerPtr) == 0) {
+        // Finished adding, retry.
+        next = tail->next_.load(std::memory_order_acquire);
+        goto retry_all;
+      }
+      if (tail_next & Node::kWakerPtr) {
+        // null next waker => list closed
+        DCHECK_NE(tail_next, Node::kWakerPtr);
+        // Node still being added, and we've already asked to be notified.
+        return Pending{};  // pending
+      }
+      DCHECK_EQ(tail_next, 0);
+      Waker* waker = new Waker(GetContext<Activity>()->MakeNonOwningWaker());
+      // Inform the adder we'd like to be woken up.
+      if (!tail->next_.compare_exchange_weak(
```

Comment:
Could you clarify why tail->next is expected to match with tail_next if tail_next is loaded from head->next_ in line 237 ?

Reply:
yikes it wasn't... think I've got this fixed